servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/AmbSingles.java

```diff
@@ -90,13 +91,22 @@ public void onSubscribe(final Cancellable cancellable) {
 
         @Override
         public void onSuccess(@Nullable final T result) {
+            propagateCancel = false;
             state.trySuccess(result);
         }
 
         @Override
         public void onError(final Throwable t) {
+            propagateCancel = false;
```

Comment:
this is what I had in mind (not tested). @idelpivnitskiy does that make sense?

```diff
diff --git a/servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/AmbSingles.java b/servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/AmbSingles.java
index 184fe6c2b..2217dfa7c 100644
--- a/servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/AmbSingles.java
+++ b/servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/AmbSingles.java
@@ -77,7 +77,14 @@ final class AmbSingles<T> extends Single<T> {
     }
 
     static final class AmbSubscriber<T> extends DelayedCancellable implements Subscriber<T> {
+        private static final int INIT = 0;
+        private static final int TERMINATED = 1;
+        private static final int CANCELLED = 2;
+        @SuppressWarnings("rawtypes")
+        private static final AtomicIntegerFieldUpdater<AmbSubscriber> doneUpdater = newUpdater(AmbSubscriber.class, "done");
+
         private final State<T> state;
+        private volatile int done;
 
         AmbSubscriber(final State<T> state) {
             this.state = state;
@@ -90,12 +97,30 @@ final class AmbSingles<T> extends Single<T> {
 
         @Override
         public void onSuccess(@Nullable final T result) {
-            state.trySuccess(result);
+            // GAS allows for terminal after cancel. Downstream knows best how to process state after cancel.
+            if (doneUpdater.getAndSet(this, TERMINATED) != TERMINATED) {
+                state.trySuccess(result);
+            }
         }
 
         @Override
         public void onError(final Throwable t) {
-            state.tryError(t);
+            // GAS allows for terminal after cancel. Downstream knows best how to process state after cancel.
+            if (doneUpdater.getAndSet(this, TERMINATED) != TERMINATED) {
+                state.tryError(t);
+            }
+        }
+
+        @Override
+        public void cancel() {
+            for (;;) {
+                final int currDone = done;
+                if (currDone != INIT) { // we could let cancelled go through here if we wanted too
+                    break;
+                } else if (doneUpdater.compareAndSet(this, INIT, CANCELLED)) {
+                    super.cancel();
+                }
+            }
         }
     }
```

Reply:
yes, this makes sense, but this is not much different from what @daschl already has in this PR and his version looks simpler. We don't need to prevent 2 terminal events at this level, we only need to prevent cancel after terminal